### PR TITLE
Ensure lambdas under target-typed contexts are always visited in nullable

### PR DESCRIPTION
**Pull request by 333fred**
We were handling target-typed conditional and switch expressions wrong in a few cases, which caused a series of issues:

1. Lambdas underneath these contexts were never visited in nullable, so warnings in their bodies weren't reported.
2. Target-typed conditionals contributed to nullable inference of the thing they were getting their target-type from in the first place.

To fix this, I've made a couple of changes. First, we add a discrete BoundConversion node for the target-typed conditional conversion. This ensures that the standard "remove conversion, visit operand, redo conversion" practice we have throughout the nullable walker handles these constructs correctly. I also split up the conversion steps for these constructs: if the construct was target typed, we now save the states for each branch at the end of their respective VisitX methods, and then when we have a target-type with nullability in VisitConversion, we pull that info out and visit the nested contexts correctly. As part of this, I fixed a couple of other bugs:

1. VisitSwitchExpression was needlessly calling VisitConversion multiple times. This caused some warnings to be double-reported, which it now only reports once.
2. We were missing IConversionOperation nodes that should have been in the IOperation tree.
3. The behaviors for GetTypeInfo and GetConversion didn't match our usual handling of explicit vs implicit conversions.

Fixes https://github.com/dotnet/roslyn/issues/51403. Fixes https://github.com/dotnet/roslyn/issues/49735.

The IOperation hook has revealed an additional bug here, that 07a0734 fixes. We were caching the governing expression for switch expressions in the switch expression binder, which caused the hook to assert for some of the newly added tests. I've removed the caching, and I'd suggest reviewing that single commit separately for ease.

### CODE REVIEW

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler @jcouv for review please.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler please review, this is a bootstrap-blocking bug. Happy to walk through the changes with anyone (including the 2 next steps I think we should take after we unblock bootstrap).

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 1]
```diff
@@ -110,12 +110,33 @@ internal partial class Binder
...
             {
-                return ConvertSwitchExpression((BoundUnconvertedSwitchExpression)source, destination, conversionIfTargetTyped: conversion, diagnostics);
+
+                var convertedSwitch = ConvertSwitchExpression((BoundUnconvertedSwitchExpression)source, destination, conversionIfTargetTyped: conversion, diagnostics);
+                return new BoundConversion(
+                    syntax,
+                    convertedSwitch,
+                    conversion,
+                    CheckOverflowAtRuntime,
+                    explicitCastInCode: isCast && !wasCompilerGenerated,
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:120]: What happens when `&& !wasCompiledGenerated` is removed from this argument?

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 2]
```diff
@@ -6991,6 +7013,60 @@ static bool dependsOnTypeParameter(TypeParameterSymbol typeParameter1, TypeParam
...
+                        (BoundExpression alternativeOperand, Conversion alternativeConversion) = RemoveConversion(alternative, includeExplicitConversions: false);
+                        var alternativeRValue = ConvertConditionalOperandOrSwitchExpressionArmResult(alternative, alternativeOperand, alternativeConversion, targetTypeWithNullability, info[1].ResultType, info[1].State, info[1].EndReachable);
+
+                        _conditionalInfoForConversionOpt.Remove(conditional);
+
+                        return consequenceRValue.State.Join(alternativeRValue.State);
+
+                    case BoundConvertedSwitchExpression { WasTargetTyped: true } @switch:
+                        Debug.Assert(_conditionalInfoForConversionOpt is not null && _conditionalInfoForConversionOpt.ContainsKey(@switch));
+                        info = _conditionalInfoForConversionOpt[@switch];
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:7039]: The reuse of a 'var' declared in a previous switch section just feels a little weird to me. Consider wrapping the sections in blocks and giving them their own scope to work with.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 3]
```diff
@@ -6991,6 +7013,60 @@ static bool dependsOnTypeParameter(TypeParameterSymbol typeParameter1, TypeParam
...
+                        return NullableFlowState.NotNull;
+
+                    case BoundUnconvertedConditionalOperator:
+                    case BoundUnconvertedSwitchExpression:
+                        return operandType.State;
+
+                    default:
+                        throw ExceptionUtilities.UnexpectedValue(conversionOperand.Kind);
+                }
+            }
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:7069]: Pausing here for the afternoon. I think I get the necessity of this change but I don't quite grasp yet what makes the dictionary necessary as opposed to keeping the additional states on the stack. Maybe we can go over it on Monday.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 4]
```diff
@@ -787,45 +787,57 @@ private void VisitSwitchExpressionCore(BoundSwitchExpression node, bool inferTyp
...
+                        var armState = getStateForArm(arm);
+                        resultTypes[i] = ConvertConditionalOperandOrSwitchExpressionArmResult(arm.Value, nodeForSyntax, conversions[i], inferredTypeWithAnnotations, resultTypes[i], armState, armState.Reachable);
+                    }
+                    inferredState = BestTypeInferrer.GetNullableState(resultTypes);
                 }
-
+            }
+            else
+            {
+                var states = ArrayBuilder<(LocalState, TypeWithState, bool)>.GetInstance();
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:817]: nit: we could pass `numSwitchArms` here.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs [hunk 5]
```diff
@@ -21,6 +21,13 @@ public override BoundNode VisitConversion(BoundConversion node)
                 return RewriteInterpolatedStringConversion(node);
             }
 
+            if (node.ConversionKind is ConversionKind.SwitchExpression or ConversionKind.ConditionalExpression)
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs:24]: I wasn't sure why it's necessary to skip these conversions in LocalRewriter. Could you please clarify?

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 6]
```diff
@@ -48219,6 +48219,243 @@ static void F(bool x, bool y, bool z, bool? w)
...
+        {
+            var comp = CreateCompilation(@"
+using System;
+class C
+{
+    static void M(bool b)
+    {
+        string? s1 = null;
+        string? s2 = """";
+        M1(s2, b ? () => s1.ToString() : () => s1?.ToString()).ToString();
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:48368]: Is it just impossible to infer `string?` here for `T` in combination with target typed switch/conditional? It feels like if the argument were `() => s1?.ToString()`, we would infer `string?` here.

**Comment by RikkiGibson** [comment]
Overall the change seems good but I'd like to make sure I understand the usage of the new state dictionary as well as make sure I understand why it's not possible for target typed operands to contribute their nullabilty to type inference.

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:120]: We violate the meaning of explicitCastInCode :). That property specifically reflects whether the user typed `(Type)`, and if the compiler generated the cast it wasn't typed by the user.

---
In reply to: [592881226](https://github.com/dotnet/roslyn/pull/51791#discussion_r592881226) [](ancestors = 592881226)

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 7]
```diff
@@ -2125,6 +2125,42 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
...
                 }
+                else if (boundExpr is BoundConvertedSwitchExpression { WasTargetTyped: true } convertedSwitch)
+                {
+                    if (highestBoundExpr is BoundConversion { ConversionKind: ConversionKind.SwitchExpression })
+                    {
+                        // There was an implicit cast.
+                        type = convertedSwitch.NaturalTypeOpt;
+                        convertedType = convertedSwitch.Type;
+                        convertedNullability = convertedSwitch.TopLevelNullability;
+                        conversion = convertedSwitch.Conversion.IsValid ? convertedSwitch.Conversion : Conversion.NoConversion;
```
[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:2136]: This is what the existing code was doing. I just moved the condition up.

---
In reply to: [593508092](https://github.com/dotnet/roslyn/pull/51791#discussion_r593508092) [](ancestors = 593508092)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs (hunk 5 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs:24]: These don't get emitted. These nodes are newly added in initial binding, they didn't exist previously.

---
In reply to: [594569914](https://github.com/dotnet/roslyn/pull/51791#discussion_r594569914) [](ancestors = 594569914)

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:120]: I'm slightly concerned about this, because synthesized code can contain explicit conversions, and NullableWalker can take such conversions and try to re-classify them, passing its `ExplicitCastInCode` value to `ClassifyConversion`'s `forCast` argument (indirectly). Then Conversions will say that the conversion doesn't exist. Wrong information could be propagated out, or unwanted diagnostics given. See #51631.

Basically, it's a rather small corner case, but I think if `ExplicitCastInCode` *requires* that a cast is present in user syntax, we may want a follow up item to make sure the NullableWalker avoids using it as an argument to `NullableWalker.GenerateConversion`, and to make sure that synthesized code never passes a 'true' value for this flag. Possibly with `Debug.Assert(!(WasCompilerGenerated && ExplicitCastInCode))` when creating a BoundConversion or something like that.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs (hunk 1 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:120]: I also feel like there is likely something I'm missing about the general purpose of `ExplicitCastInCode`, something that is not addressed by just `Conversion.IsExplicit`.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:120]: > something that is not addressed by just Conversion.IsExplicit

Whether or not a conversion is explicit is unrelated to whether it actually existed in code. You can explicitly specify an implicit conversion, or the only conversion that exists could be explicit and not be specified in source (which would lead to an error). I think any concerns here are beyond the scope of the changes in this method, feel free to file a bug for follow up later. Note this is how we calculate `ExplicitCastInCode` in all cases in this method, it's not something I'm coming up with here.

---
In reply to: [594763328](https://github.com/dotnet/roslyn/pull/51791#discussion_r594763328) [](ancestors = 594763328)

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:120]: Have filed #51896, thanks.

File: src/Compilers/CSharp/Portable/Binder/SwitchExpressionBinder.cs [hunk 8]
```diff
@@ -176,59 +166,32 @@ private TypeSymbol InferResultType(ImmutableArray<BoundSwitchExpressionArm> swit
...
-
-        private void EnsureSwitchGoverningExpressionAndDiagnosticsBound()
+        internal (TypeSymbol GoverningType, uint GoverningValEscape) GetInputTypeAndValEscape(BoundExpression? inputExpression = null)
         {
-            if (_inputExpression == null)
-            {
-                var switchGoverningDiagnostics = new BindingDiagnosticBag();
-                var boundSwitchGoverningExpression = BindSwitchGoverningExpression(switchGoverningDiagnostics);
-                _inputExpressionDiagnostics = switchGoverningDiagnostics;
-                Interlocked.CompareExchange(ref _inputExpression, boundSwitchGoverningExpression, null);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/SwitchExpressionBinder.cs:224]: > The IOperation hook has revealed an additional bug here, that 07a0734 fixes. We were caching the governing expression for switch expressions in the switch expression binder, which caused the hook to assert for some of the newly added tests. I've removed the caching, and I'd suggest reviewing that single commit separately for ease.

I didn't understand how the caching of the bound governing expression resulted in a problem with the IOperation hook.

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs (hunk 5 above)
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs:24]: Was this change necessary? For target-typed `new()`, I think the expression is also wrapped in a `BoundConversion` during initial binding, but we don't erase that conversion during lowering.

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs [hunk 9]
```diff
@@ -110,12 +110,33 @@ internal partial class Binder
 
             if (conversion.Kind == ConversionKind.SwitchExpression)
             {
-                return ConvertSwitchExpression((BoundUnconvertedSwitchExpression)source, destination, conversionIfTargetTyped: conversion, diagnostics);
+
+                var convertedSwitch = ConvertSwitchExpression((BoundUnconvertedSwitchExpression)source, destination, conversionIfTargetTyped: conversion, diagnostics);
+                return new BoundConversion(
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:115]: Can we move this `new BoundConversion` into `ConvertSwitchExpression`? (would be conditional on having a `conversionIfTargetTyped`)
I think that would help reduce sprawl in the current method.

Same for `ConvertConditionalExpression` below.

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs [hunk 10]
```diff
@@ -787,45 +787,57 @@ private void VisitSwitchExpressionCore(BoundSwitchExpression node, bool inferTyp
...
                     var nodeForSyntax = expressions[i];
-                    var conversionOpt = node.SwitchArms[i].Value switch { BoundConversion c when c != nodeForSyntax => c, _ => null };
-                    // Report top-level warnings
-                    _ = VisitConversion(conversionOpt, conversionOperand: nodeForSyntax, conversions[i], targetTypeWithNullability: inferredTypeWithAnnotations, operandType: resultTypes[i],
-                        checkConversion: true, fromExplicitCast: false, useLegacyWarnings: false, AssignmentKind.Assignment, reportRemainingWarnings: false, reportTopLevelWarnings: true);
+                    var armState = getStateForArm(node.SwitchArms[i]);
+                    states.Add((armState, resultTypes[i], armState.Reachable));
                 }
+
+                _conditionalInfoForConversionOpt ??= PooledDictionary<BoundExpression, ImmutableArray<(LocalState, TypeWithState, bool)>>.GetInstance();
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs:825]: >_conditionalInfoForConversionOpt ??= PooledDictionary<BoundExpression, ImmutableArray<(LocalState, TypeWithState, bool)>>.GetInstance(); [](start = 16, length = 136)

nit: Consider extracting as `ConditionalInfoForConversion` here and in `VisitConditionalOperatorCore`. Something like `ConditionalInfoForConversion.Add(node, states.ToImmutableAndFree());`

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 11]
```diff
@@ -4355,47 +4364,54 @@ private static TypeWithState GetNullCoalescingResultType(TypeWithState rightResu
...
-                    resultTypeWithAnnotations,
-                    alternativeRValue,
-                    alternativeState,
-                    alternativeEndReachable);
+                resultState = consequenceRValue.State.Join(alternativeRValue.State);
 
-                resultState = convertedConsequenceResult.State.Join(convertedAlternativeResult.State);
+                _conditionalInfoForConversionOpt ??= PooledDictionary<BoundExpression, ImmutableArray<(LocalState, TypeWithState, bool)>>.GetInstance();
+                _conditionalInfoForConversionOpt.Add(node, ImmutableArray.Create(
+                    (consequenceState, consequenceRValue, consequenceEndReachable),
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:4413]: >consequenceState [](start = 21, length = 16)

I think `consequenceState` is the state *after* visiting the consequence when we reach here. But I expected that we'd store the state *before* visiting the consequence, based on the comment on `_conditionalInfoForConversionOpt`.

> This is then used by VisitConversion to properly set the state *before* each branch when visiting a conversion applied to such a construct.

```
// line 4334 above
                (consequence, consequenceConversion, consequenceRValue) = visitConditionalOperand(consequenceState, originalConsequence);
                Unsplit();
                consequenceState = this.State;
```

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs [hunk 12]
```diff
@@ -4355,47 +4364,54 @@ private static TypeWithState GetNullCoalescingResultType(TypeWithState rightResu
...
+                        resultTypeWithAnnotations,
+                        alternativeRValue,
+                        alternativeState,
+                        alternativeEndReachable);
+
+                    resultState = convertedConsequenceResult.State.Join(convertedAlternativeResult.State);
                 }
             }
             else
             {
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:4408]: If I understood correctly, the strategy is to postpone/delay analysis of the branch when visiting the target-typed conditional/switch expression, but come back to it when we visit the conversion that provides a target-type.
We do this by storing information from `VisitConditionalOperator` and passing it up to `VisitConversion`, so that we can restore the proper state before each branch.

nit: if that's correct, it may be good to add a comment.

Would it be possible for `VisitConversion` to pass information (namely the target type) down to `VisitConditionalOperator` instead? If possible, that would seem simpler and less information to store.

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 6). Will look at the test changes tomorrow.

File: src/Compilers/CSharp/Portable/Binder/SwitchExpressionBinder.cs (hunk 8 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/SwitchExpressionBinder.cs:224]: > I didn't understand how the caching of the bound governing expression resulted in a problem with the IOperation hook. 

Because the node is cached, subsequent bindings will see the same node. It has already had its top-level nullability set, and the DebugVerifier asserts because it should only be seeing new nodes, not previously-analyzed nodes.

---
In reply to: [595616905](https://github.com/dotnet/roslyn/pull/51791#discussion_r595616905) [](ancestors = 595616905)

File: src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs (hunk 5 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Conversion.cs:24]: Yes, it's necessary. Emit asserts if this isn't removed.

---
In reply to: [595617051](https://github.com/dotnet/roslyn/pull/51791#discussion_r595617051) [](ancestors = 595617051)

File: src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs (hunk 9 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs:115]: I intentionally didn't do that, because `ConvertSwitchExpression` is used in other places.

---
In reply to: [595617069](https://github.com/dotnet/roslyn/pull/51791#discussion_r595617069) [](ancestors = 595617069)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 11 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:4413]: It depends. If the nested operator was a lambda, it's unchanged from before the visit operand, because the lambda was skipped. If it isn't a lambda, then it's the state after visiting, which is still correct because that's the state that should be set when we call `VisitConversion` for real on the operand.

---
In reply to: [595627389](https://github.com/dotnet/roslyn/pull/51791#discussion_r595627389) [](ancestors = 595627389)

File: src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs (hunk 12 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:4408]: > Would it be possible for VisitConversion to pass information (namely the target type) down to VisitConditionalOperator instead? If possible, that would seem simpler and less information to store, and maybe we can factor with code calling ConvertConditionalOperandOrSwitchExpressionArmResult above.

No, or more precisely yes, but _everywhere_ that one of these things could be would need to have a ton of special code added. Everywhere in the walker that we call `RemoveConversion` would need to be updated to specially detect if the operand was target-typed, and if so do something special. That's a lot of peanut butter spread across the walker and I'm not certain that we'd catch everything. This has the advantage of working with our existing pattern.

> the strategy is to postpone/delay analysis of the branch when visiting the target-typed conditional/switch expression

Not quite. We delay analysis of the _conversion_ on the branch. Not on the branch itself.

---
In reply to: [595629506](https://github.com/dotnet/roslyn/pull/51791#discussion_r595629506) [](ancestors = 595629506)

Comment (333fred) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:4408]: I'll add some more comments to try and convey this.

---
In reply to: [597050103](https://github.com/dotnet/roslyn/pull/51791#discussion_r597050103) [](ancestors = 597050103,595629506)

Comment (jcouv) [reply src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs:4408]: Thanks. Additional comments will be useful

---
In reply to: [597050322](https://github.com/dotnet/roslyn/pull/51791#discussion_r597050322) [](ancestors = 597050322,597050103,595629506)

**Comment by 333fred** [comment]
@jcouv addressed feedback, please take another look.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 7)

**Comment by 333fred** [comment]
I've done a rebase to condense the changes into their respective parts:

1. The lambda fix
2. The switch governing expression fix.
3. Enable the bootstrap build

That last commit is the only new thing, let's see if that passes.

**Review by RikkiGibson — APPROVED** [review]

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 4)
